bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/PJLinkDeviceHandler.java

```diff
@@ -0,0 +1,329 @@
+import java.util.stream.Collectors;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.core.validation.ConfigValidationException;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.StateOption;
+import org.openhab.binding.pjlinkdevice.internal.device.PJLinkDevice;
+import org.openhab.binding.pjlinkdevice.internal.device.command.AuthenticationException;
+import org.openhab.binding.pjlinkdevice.internal.device.command.ResponseException;
+import org.openhab.binding.pjlinkdevice.internal.device.command.errorstatus.ErrorStatusQueryResponse.ErrorStatusDevicePart;
+import org.openhab.binding.pjlinkdevice.internal.device.command.errorstatus.ErrorStatusQueryResponse.ErrorStatusQueryResponseState;
+import org.openhab.binding.pjlinkdevice.internal.device.command.input.Input;
+import org.openhab.binding.pjlinkdevice.internal.device.command.mute.MuteInstructionCommand.MuteInstructionChannel;
+import org.openhab.binding.pjlinkdevice.internal.device.command.mute.MuteQueryResponse.MuteQueryResponseValue;
+import org.openhab.binding.pjlinkdevice.internal.device.command.power.PowerQueryResponse.PowerQueryResponseValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+
+/**
+ * The {@link PJLinkDeviceHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Nils Schnabel - Initial contribution
+ */
+@NonNullByDefault
+public class PJLinkDeviceHandler extends BaseThingHandler {
+
+  private @Nullable PJLinkDeviceConfiguration config;
+
+  private @Nullable PJLinkDevice device;
+
+  private InputChannelStateDescriptionProvider stateDescriptionProvider;
+
+  private final Logger logger = LoggerFactory.getLogger(PJLinkDeviceHandler.class);
+
+  private @Nullable ScheduledFuture<?> refreshJob;
+
+  private @Nullable ScheduledFuture<?> setupJob;
+
+  public PJLinkDeviceHandler(Thing thing, InputChannelStateDescriptionProvider stateDescriptionProvider) {
+    super(thing);
+    this.stateDescriptionProvider = stateDescriptionProvider;
+  }
+
+  @Override
+  public void dispose() {
+    clearSetupJob();
+
+    clearRefreshInterval();
+
+    PJLinkDevice device = this.device;
+    if (device != null) {
+      device.dispose();
+    }
+
+    this.config = null;
+    this.device = null;
+  }
+
+  public void refresh(PJLinkDeviceConfiguration config) {
```

Comment:
Why do you need config as a parameter here? I would guess that the configuration is not changing at all, so you could just acces the class field.

Reply:
Directly accessing the class field raises nullability warnings (class field might not have been set yet). 

In case it is not yet set, I'm risking ConfigurationExceptions during creation of one during getConfiguration(). As those would occur in a job, exception handling would be non-trivial.

So in the end, it's easier to give the job what it needs to run as a parameter.

Reply:
why should that be null if you initialuite the config before in `initialize()`? If you do so, you can add `@NonNullByDefault({})`to the config field.